system/security/plugins/jwtSecurity/jwtSecurity.cpp

```diff
@@ -0,0 +1,1159 @@
+            }
+
+            // Add file scope permissions in alphabetical (and implicitly, length) order
+            std::sort(fileScopePerms.begin(), fileScopePerms.end());
+            for (CollectedScopePerms::const_iterator x = fileScopePerms.begin(); x != fileScopePerms.end(); x++)
+            {
+                if (x->claim == "AllowFileScopeAccess")
+                    userInfo->addFileScopePerm(x->scopePattern, SecAccess_Access, defaultFileAccess);
+                else if (x->claim == "AllowFileScopeRead")
+                    userInfo->addFileScopePerm(x->scopePattern, SecAccess_Read, defaultFileAccess);
+                else if (x->claim == "AllowFileScopeWrite")
+                    userInfo->addFileScopePerm(x->scopePattern, SecAccess_Write, defaultFileAccess);
+                else if (x->claim == "AllowFileScopeFull")
+                    userInfo->addFileScopePerm(x->scopePattern, SecAccess_Full, defaultFileAccess);
+                else if (x->claim == "DenyFileScopeAccess")
+                    userInfo->addFileScopePerm(x->scopePattern, SecAccess_Access, defaultFileAccess, true);
+                else if (x->claim == "DenyFileScopeRead")
+                    userInfo->addFileScopePerm(x->scopePattern, SecAccess_Read, defaultFileAccess, true);
+                else if (x->claim == "DenyFileScopeWrite")
+                    userInfo->addFileScopePerm(x->scopePattern, SecAccess_Write, defaultFileAccess, true);
+                else if (x->claim == "DenyFileScopeFull")
+                    userInfo->addFileScopePerm(x->scopePattern, SecAccess_Full, defaultFileAccess, true);
+            }
+
+            gUserCache.set(subject, userInfo);
+
+            verificationResult = true;
+        }
+        catch (const std::runtime_error& e)
+        {
+            // Most jwt-cpp exceptions are subclasses of std::runtime_error
+            PROGLOG("CJwtSecurityManager::verifyToken error: %s", e.what());
+        }
+        catch (const std::logic_error& e)
+        {
+            PROGLOG("CJwtSecurityManager::logic_error: %s", e.what());
+        }
+        catch (const std::bad_cast& e)
+        {
+            PROGLOG("CJwtSecurityManager::bad_cast error: %s", e.what());
+        }
+        catch (...)
+        {
+            PROGLOG("CJwtSecurityManager::verifyToken unknown error");
+            throw;
+        }
+
+        return verificationResult;
+    }
+
+    /**
+     * Load the contents of the key if necessary.  Contents loaded when
+     * at least KEY_FILE_CHECK_INTERVAL have passed since last check time
+     *
+     * Function will throw an exception on error.
+     *
+     * @return  true
+     */
+    bool ensureKeyLoaded()
+    {
+        if ((keyContentsTimer.elapsedMs() > KEY_FILE_CHECK_INTERVAL_MS) || keyContents.empty())
+        {
+            CriticalBlock block(crit);
+
+            if ((keyContentsTimer.elapsedMs() > KEY_FILE_CHECK_INTERVAL_MS) || keyContents.empty())
```

Comment:
Two issues.
- if keyContents is only used from verifyToken, it would be clearer if the call to ensureKeyLoaded was only in that function, and there  was no critical block in this funtion.
- repeating a test after entering a critical block isn't technically thread safe in c++ (unless they are atomic variables, or there is a memory barrier).  In this case I think the test keyContents.empty() could fail to be re-evaluated!  (Nasty - see https://www.aristeia.com/Papers/DDJ_Jul_Aug_2004_revised.pdf)

Reply:
Excellent observation re: keyContents.  I've changed things so that ensureKeyLoaded() is called only from within verifyToken(), which is also the only place in which that member is read, then protected both in the same critical block.